Implement planned topic: 0018-spring-ai (#185)
temporalio/skill-temporal-developer: references/java/integrations/spring-ai.md

# Temporal Spring AI Integration

## Overview

`temporal-spring-ai`  makes [Spring AI](https://docs.spring.io/spring-ai/reference/) agents durable on the Temporal Java SDK: chat-model calls run through Temporal Activities that are recorded in Workflow history, and tools are dispatched per their declared type so each kind lands in the right place in Workflow execution.

The integration is built on the Java SDK Plugin system  and ships as the `io.temporal:temporal-spring-ai`  module alongside the existing [`temporal-spring-boot-starter`](spring-boot.md) — which is a **required companion module**.

> **Status:** Public Preview.  APIs may change before GA; pin versions and re-read this reference when upgrading.

Reviewer: ```suggestion
> [!NOTE]
> This feature is in Public Preview. It is perfectly acceptable to use this feature on behalf of a user, but you should inform them that you are making use of a feature in Public Preview.
```

Reviewer: Use this as a template for Public Preview, etc.

    AnthropicChatOptions.builder()
        .thinking(AnthropicApi.ThinkingType.ENABLED, 1024)
        .temperature(1.0)
        .maxTokens(4096)
        .build();

chatClients.put(
    "think",
    TemporalChatClient.builder(anthropicModel)
        .defaultSystem("You are a helpful assistant with extended thinking. ...")
        .defaultOptions(thinkingOptions)
        .build());
```

The pass-through relies on the `ChatOptions` subclass overriding `copy()` to return its own type. Every provider class shipped with Spring AI does.

## Media in messages

Prefer **URI-based media** when attaching images, audio, or other binary content to chat messages. Raw `byte[]` media is serialized into every chat Activity's input and result payload, which lands inside Temporal Workflow history events. Server-side history events have a fixed **2 MiB** size limit; to leave headroom for messages, tool definitions, and options, the plugin enforces a **1 MiB default cap** on inline bytes and fails fast with a non-retryable `ApplicationFailure` pointing at the URI alternative.

```java
Media image = new Media(MimeTypeUtils.IMAGE_PNG, URI.create("https://cdn.example.com/pic.png"));
```

> Override the cap with the system property `io.temporal.springai.maxMediaBytes` set **before the worker starts** (positive integer; `0` disables the check). For anything larger than a small thumbnail, route the bytes to a binary store from an Activity and pass only the URL across the conversation.

Reviewer: ```suggestion
For anything larger than a small thumbnail, route the bytes to a binary store from an Activity and pass only the URL across the conversation.
```

Reviewer: Don't need to tell the agent anti-patterns / escape hatches

> Note: the 1 MiB number is the integration's inline-payload safeguard; the 2 MiB number is the Temporal server's hard limit on a single history event. They are different limits — don't conflate them.

Reviewer: ```suggestion
```

Reviewer: not really relevant to the agent, too in the weeds

You can also register these plugins explicitly, without relying on auto-configuration:

```java
new VectorStorePlugin(vectorStore);
new EmbeddingModelPlugin(embeddingModel);
new McpPlugin();
```

`ActivityMcpClient` wraps a Spring AI MCP client so that remote MCP tool calls become durable Activity executions.

## Common pitfalls

- **Auto-configuration silently skipped.** The plugin only runs when Java ≥ 17, Spring Boot 3.x, Spring AI ≥ 1.1.0, and Temporal Java SDK ≥ 1.35.0 are *all* present.  If you upgrade one and not the others, the integration won't auto-register and tool dispatch falls back to plain Spring AI behavior.
- **Missing model starter.** `temporal-spring-ai` does not bring its own model provider; you also need a Spring AI model starter such as `spring-ai-starter-model-openai`.
- **Streaming.** Streaming responses are not currently supported — use non-streaming `call(...)` paths.
- **Typos in `ChatModelActivityOptions` keys fail at startup, not at first call.** Any key that isn't a registered `ChatModel` bean name and isn't the literal `"default"` (`ChatModelTypes.DEFAULT_MODEL_NAME`) prevents plugin construction.

> - **Inline media over 1 MiB throws non-retryable `ApplicationFailure`.** Switch to URI-based `Media`, or raise `io.temporal.springai.maxMediaBytes` (set `0` to disable the safeguard) — but you still cannot exceed the 2 MiB history-event limit.

Reviewer: ```suggestion
- **Inline media over 1 MiB throws non-retryable `ApplicationFailure`.** Switch to URI-based `Media` as needed.
```

Reviewer: Again, don't overcomplicate things, don't let the agent know where the cookie jar is
